core/src/sumeragi/mod.rs

```diff
@@ -127,22 +143,76 @@ impl Sumeragi {
             .internal
             .wsv
             .lock()
-            .expect("Failed to lock on `update_metrics`. Mutex poisoned");
+            .expect("Failed to lock on `wsv`. Mutex poisoned");
+
+        let metrics_guard = self
+            .metrics_mutex
+            .lock()
+            .expect("Failed to lock on `metrics`. Mutex poisoned");
+
+        let mut last_block_height_guard = self
+            .last_block_height
+            .lock()
+            .expect("Failed to lock on `last_block_height_when_updating_metrics`. Mutex poisoned");
+
+        let start_index = *last_block_height_guard;
+        {
+            let blocks_iter = wsv_guard.blocks();
+            let blocks_iter =
+                blocks_iter.skip(start_index.try_into().expect("Failed to cast to u32."));
+            for block in blocks_iter {
+                let block_txs_accepted = block.as_v1().transactions.len() as u64;
+                let block_txs_rejected = block.as_v1().rejected_transactions.len() as u64;
+
+                metrics_guard
+                    .txs
+                    .with_label_values(&["accepted"])
+                    .inc_by(block_txs_accepted);
+                metrics_guard
+                    .txs
+                    .with_label_values(&["rejected"])
+                    .inc_by(block_txs_rejected);
+                metrics_guard
+                    .txs
+                    .with_label_values(&["total"])
+                    .inc_by(block_txs_accepted + block_txs_rejected);
+                metrics_guard.block_height.inc();
+            }
+            *last_block_height_guard = wsv_guard.height();
+        }
+
+        metrics_guard.domains.set(wsv_guard.domains().len() as u64);
+
+        let mut last_metric_tx_amounts_guard = self
+            .last_metric_tx_amounts
+            .lock()
+            .expect("Failed to lock on `last_metric_tx_amounts`. Mutex poisoned");
+        let mut last_metric_tx_amounts_counter_guard = self.last_metric_tx_amounts_counter.lock().expect("Failed to lock on `last_total_tx_amounts_counter_when_updating_metrics`. Mutex poisoned");
+
+        let diff_count =
+            wsv_guard.metric_tx_amounts_counter.get() - *last_metric_tx_amounts_counter_guard;
+        let diff_amount_per_count = (wsv_guard.metric_tx_amounts.get()
+            - *last_metric_tx_amounts_guard)
+            / (diff_count as f64);
+        for _ in 0..diff_count {
+            *last_metric_tx_amounts_counter_guard += 1;
+            *last_metric_tx_amounts_guard += diff_amount_per_count;
+
+            metrics_guard.tx_amounts.observe(diff_amount_per_count);
```

Comment:
Also a bit confused about the reasoning for this refactor here, and it seems that we're losing a bit of precision? Because instead of directly adding observations to the `Histogram` on each mint/burn/transfer like before, we aggregate them and then average out the `tx_amounts` difference across the whole window, instead of recording it precisely. Could you explain the purpose here?

Reply:
Internally `Histogram` does the same thing I am doing. It stores a total amount and the amount of observations. Except for some precision lost in the floating point maths it is the exact same value.

I would like to improve the precision of the gathered metrics but that was outside the scope of this PR. Since this PR is simply about decoupling the metrics from the WSV.

Reply:
> Internally Histogram does the same thing I am doing. It stores a total amount and the amount of observations. Except for some precision lost in the floating point maths it is the exact same value.

Histograms always do that. The more precise approximation is called a kernel-density estimate, and we have neither the need, nor the time to implement it in Rust. An even more precise approach would be to return all of the transaction amounts given an immutable view of a blockchain, but I don't think I need to explain why this is a bad idea. 

> I would like to improve the precision of the gathered metrics but that was outside the scope of this PR. Since this PR is simply about decoupling the metrics from the WSV.

I don't think that's necessary for the time being. We might revisit this, once metrics are used in full force.